### PR TITLE
[Feature] Add RandomAffine transform.

### PR DESCRIPTION
**Pull request by RangiLyu**
Add RandomAffine transform which is used in YOLOX.

### CODE REVIEW

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+                                 0.5 + self.translate_ratio) * width
+        trans_y = random.uniform(0.5 - self.translate_ratio,
+                                 0.5 + self.translate_ratio) * height
+        translate_matrix = self._get_translation_matrix(trans_x, trans_y)
+
+        warp_matrix = (
+            translate_matrix @ shear_matrix @ rotation_matrix @ scaling_matrix
+            @ center_matrix)
+
+        img = cv2.warpPerspective(
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:1999]: This is Perspective, not affine. Maybe we should support at the same time.

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+    This operation randomly generates affine transform matrix which including
+    rotation, translation, shear and scaling transforms.
+
+    Args:
+        rotate_degrees (float): Maximum degrees of rotation transform.
+            Default: 10.
+        translate_ratio (float): Maximum ratio of translation. Default: 0.1.
+        scaling_ratio (tuple[float]): Min and max ratio of scaling transform.
+            Default: (0.5, 1.5).
+        shear_degrees (float): Maximum degrees of shear transform. Default: 2.
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:1923]: shear_degrees shear_degree

File: mmdet/datasets/pipelines/transforms.py (hunk 1 above)
Comment (RangiLyu) [reply mmdet/datasets/pipelines/transforms.py:1999]: When the third row of the homography matrix is [0,0,1], warpPerspective is equivalent to warpAffine mathematically.

Comment (hhaAndroid) [reply mmdet/datasets/pipelines/transforms.py:1999]: But `warpPerspective` and `RandomAffine` are inconsistent.

File: mmdet/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+        translate_ratio (float): Maximum ratio of translation. Default: 0.1.
+        scaling_ratio (tuple[float]): Min and max ratio of scaling transform.
+            Default: (0.5, 1.5).
+        shear_degree (float): Maximum degrees of shear transform. Default: 2.
+        border (tuple[int]): Distance from height and width sides of input
+            image to adjust output shape. Only used in mosaic dataset.
+            Default: (0, 0).
+        border_val (tuple[int]): Border padding values of 3 channels.
+            Default: (114, 114, 114).
+        wh_filter_thr (bool): Width and height threshold to filter bboxes.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1929]: The type should be float

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+            If the height or width of a box is smaller than this value, it
+            will be removed. Default: 2.
+        area_ratio_filter_thr (tuple): Threshold of area ratio between
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1932]: float?

File: mmdet/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+            original bboxes and wrapped bboxes. If smaller than this value,
+            the box will be removed. Default: 0.2.
+        aspect_ratio_filter_thr (int): Aspect ratio of width and height
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1935]: float?

File: mmdet/datasets/pipelines/transforms.py [hunk 6]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+    """Random affine transform data augmentation.
+
+    This operation randomly generates affine transform matrix which including
+    rotation, translation, shear and scaling transforms.
+
+    Args:
+        rotate_degree (float): Maximum degrees of rotation transform.
+            Default: 10.
+        translate_ratio (float): Maximum ratio of translation. Default: 0.1.
+        scaling_ratio (tuple[float]): Min and max ratio of scaling transform.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1921]: `scaling_ratio` sounds like it is a float number
Can we rename it to scaling_ratio_range

**Review by jshilong — APPROVED** [review]
LGTM

File: mmdet/datasets/pipelines/transforms.py [hunk 7]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+
+@PIPELINES.register_module()
+class RandomAffine:
+    """Random affine transform data augmentation.
+
+    This operation randomly generates affine transform matrix which including
+    rotation, translation, shear and scaling transforms.
+
+    Args:
+        rotate_degree (float): Maximum degrees of rotation transform.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1918]: Can we add `max` prefix for `rotate_degree` and `translate_ratio`，`shear_degree` ？

File: mmdet/datasets/pipelines/transforms.py [hunk 8]
```diff
@@ -1903,3 +1905,195 @@ def __repr__(self):
...
+        translate_ratio (float): Maximum ratio of translation. Default: 0.1.
+        scaling_ratio (tuple[float]): Min and max ratio of scaling transform.
+            Default: (0.5, 1.5).
+        shear_degree (float): Maximum degrees of shear transform. Default: 2.
+        border (tuple[int]): Distance from height and width sides of input
+            image to adjust output shape. Only used in mosaic dataset.
+            Default: (0, 0).
+        border_val (tuple[int]): Border padding values of 3 channels.
+            Default: (114, 114, 114).
+        wh_filter_thr (float): Width and height threshold to filter bboxes.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1929]: Seem we always use `min` and `max` as prefix when representing the threshold of the shape of bbox and image, such as `max_shape` in bboxcoder and `min_bbox_size` in RPN, seem `thr` only used when it is related to score

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**